File: DSA /Unit 05/Lec 30/Q3.cpp
Single File Programming Question
Problem Statement



Karthik is arranging a set of numbers using a Min Heap so that the smallest values always rise to the top. After inserting all the elements, he wants to view the complete heap and also print the first k smallest elements from it. 



Help him display the final heap structure and the first k elements.

Input format :
The first line contains two integers N and k, where N is the number of elements and k is the count of smallest elements to print.

The second line contains N space-separated integers, representing the values to be inserted into the Min Heap.

Output format :
The first line prints the final Min Heap after all insertions.

The second line prints the first k elements from the heap array.



Refer to the sample output for formatting specifications.

Code constraints :
The given test cases fall under the following constraints:

1 ≤ N ≤ 100

1 ≤ k ≤ N

Sample test cases :
Input 1 :
7 3
12 4 19 7 2 15 9
Output 1 :
2 4 9 12 7 19 15 
2 4 9 
Input 2 :
5 2
8 3 10 1 6
Output 2 :
1 3 10 8 6 
1 3 

#include <iostream>
using namespace std;

void heapifyUp(int heap[], int index) {
    while (index > 0) {
        int parent = (index - 1) / 2;
        if (heap[parent] > heap[index]) {
            swap(heap[parent], heap[index]);
            index = parent;
        } else {
            break;
        }
    }
}

int main() {
    int N, k;
    cin >> N >> k;

    int heap[100];

    for (int i = 0; i < N; i++) {
        cin >> heap[i];
        heapifyUp(heap, i);
    }

    for (int i = 0; i < N; i++) {
        cout << heap[i] << " ";
    }
    cout << endl;

    for (int i = 0; i < k; i++) {
        cout << heap[i] << " ";
    }

    return 0;
}

File: Java/Unit 04/Lec 22/Q3.java
Single File Programming Question
Problem Statement



Sharon, a fitness enthusiast, is working on a project to track and display the improvement percentage in weightlifting for gym members. 
  She needs to create a GymMember class with a non-static nested class Progress that calculates the percentage improvement in weightlifting over time. 



The program should take input from gym members, including their names, initial weight, and current weight, and then calculate and display the improvement percentage. 



Formula

Improvement Percentage = (Current Weight − Initial Weight) / Initial Weight * 100

Input format :
The first input line consists of a string, representing the person's name.

The second line consists of two space-separated integers, representing the initial weight and current weight of the person, respectively.

Output format :
The output prints a double value rounded off to one decimal place, representing the improvement percentage.



Refer to the sample output for formatting specifications.

Code constraints :
In this scenario, the test cases fall under the following constraints:

1 ≤ weight ≤ 100

Sample test cases :
Input 1 :
John
50 55
Output 1 :
10.0
Input 2 :
Dev
60 68
Output 2 :
13.3
Input 3 :
Ragu
60 98
Output 3 :
63.3



import java.util.Scanner;

// Solution class
class Solution {

    // Outer class
    class GymMember {
        private String name;
        private int initialWeight;
        private int currentWeight;

        // Constructor
        public GymMember(String name, int initialWeight, int currentWeight) {
            this.name = name;
            this.initialWeight = initialWeight;
            this.currentWeight = currentWeight;
        }

        // Non-static nested class
        class Progress {

            // Method to calculate improvement percentage
            public double calculate() {
                return ((currentWeight - initialWeight) * 100.0) / initialWeight;
            }
        }

        // Method called from main
        public double getImprovementPercentage() {
            Progress p = new Progress();   // create inner class object
            return p.calculate();         // return result
        }
    }
}

// Footer (given)
class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        String name = scanner.nextLine();
        int initialWeight = scanner.nextInt();
        int currentWeight = scanner.nextInt();

        Solution solution = new Solution();
        Solution.GymMember member =
                solution.new GymMember(name, initialWeight, currentWeight);

        System.out.printf("%.1f\n", member.getImprovementPercentage());

        scanner.close();
    }
}
